example_updates/rollback_update/example-rollback-update.sh

```diff
@@ -0,0 +1,798 @@
+            extendedResultCode=0
+            resultDetails="Already installed."
+            ret_val=0
+        fi
+    fi
+   
+
+    # Prepare ADUC_Result json.
+    make_aduc_result_json "$resultCode" "$extendedResultCode" "$resultDetails" aduc_result
+
+    # Show output.
+    output "Result:" "$aduc_result"
+
+    # Write ADUC_Result to result file.
+    result "$aduc_result"
+
+    $ret $ret_val
+}
+
+#
+# Performs steps to finalize the update.
+#
+ApplyUpdate() {
+    log_info "Applying update..."
+
+    ret_val=0
+    resultCode=0
+    extendedResultCode=0
+    resultDetails=""
+
+    curr_workflow=$(basename $workfolder)
+
+    echo "Applying." >> "${log_file}"
+
+    # Set the bootloader environment variable
+    # to tell the bootloader to boot into the new partition.
+    #
+    # 'rpipart' variable is specific to our boot.scr script.
+    
+    fw_setenv rpipart $update_partition
+
+    # Set the updated flag and set the past_workflow to the current workflow
+    fw_setenv have_updated 1
+    fw_setenv past_workflow "$curr_workflow"
+    fw_setenv boot_attempts 0
+    ret_val=$?
+
+    if [[ $ret_val -eq 0 ]]; then
+
+        log_info "restart_to_apply=$restart_to_apply"
+        log_info "restart_agent_to_apply=$restart_agent_to_apply"
+
+        if [[ ${restart_to_apply} == "yes" ]]; then
+            log_info "Returning ADUC_Result_Apply_RequiredImmediateReboot(705)"
+            resultCode=705
+        elif [[ ${restart_agent_to_apply} == "yes" ]]; then
+            log_info "Retrurning ADUC_Result_Apply_RequiredImmediateAgentRestart(707)"
+            resultCode=707
+        else
+            log_info "Returning ADUC_Result_Apply_Success(700)"
+            resultCode=700
+        fi
+        extendedResultCode=0
+        resultDetails=""
+    else
+        resultCode=0
+        make_swupdate_handler_erc "$ret_val" extendedResultCode
+        resultDetails="Cannot set rpipart value to $update_partition"
+        log_error "$resultDetails"
+    fi
+
+    # Prepare ADUC_Result json.
+    aduc_result=""
+    make_aduc_result_json "$resultCode" "$extendedResultCode" "$resultDetails" aduc_result
+
+    # Show output.
+    output "Result:" "$aduc_result"
+
+    # Write ADUC_Result to result file.
+    result "$aduc_result"
+
+    $ret $ret_val
+}
+
+#
+# Cancel current update.
+#
+# Set the bootloader environment variable to tell the bootloader to boot into the current partition
+# instead of the one that was updated. Note: rpipart variable is specific to our boot.scr script.
+#
+CancelUpdate() {
+    log_info "CancelUpdate called"
+
+    # Set the bootloader environment variable
+    # to tell the bootloader to boot into the current partition.
+    # rpipart variable is specific to our boot.scr script.
+    echo "Cancelling update." >> "${log_file}"
+    resultCode=0
+    extendedResultCode=0
+    resultDetails=""
+    ret_val=
+
+    echo "Revert update." >> "${log_file}"
+    fw_setenv rpipart $current_partition
```

Comment:
in gen1 agent, cancel can come in asynchronously, so technically there's a race condition here--might not be worth it to do now but could consider things like flock for lock file and waiting until lock is relinquished before continuing, etc.